test/parallel/test-http2-client-destroy.js

```diff
@@ -71,19 +71,40 @@ const h2 = require('http2');
   );
 }
 
-// test destroy before connect
+// test destroy before client operations
 {
   const server = h2.createServer();
   server.listen(
     0,
     common.mustCall(() => {
       const client = h2.connect(`http://localhost:${server.address().port}`);
-
-      const req = client.request({ ':path': '/' });
+      const req = client.request();
       client.destroy();
 
       req.on('response', common.mustNotCall());
       req.resume();
+
+      const sessionError = {
+        type: Error,
+        code: 'ERR_HTTP2_INVALID_SESSION',
+        message: 'The session has been destroyed'
+      };
+
+      common.expectsError(() => client.request(), sessionError);
+      common.expectsError(() => client.settings({}), sessionError);
+      common.expectsError(() => client.priority(req, {}), sessionError);
+      common.expectsError(() => client.shutdown(), sessionError);
+
+      // Wait for setImmediate call from destroy() to complete
+      // so that state.destroyed is set to true
+      setTimeout(() => {
+        common.expectsError(() => client.request(), sessionError);
+        common.expectsError(() => client.settings({}), sessionError);
+        common.expectsError(() => client.priority(req, {}), sessionError);
+        common.expectsError(() => client.shutdown(), sessionError);
+        common.expectsError(() => client.rstStream(req), sessionError);
+      }, 10);
```

Comment:
This should be `common.platformTimeout(10)` — I would also strongly recommend trying to find a way to avoid the timeout. In general we try to avoid using timeout unless it's part of what's being tested.

(I haven't looked at the code so I don't know if it's possible. Perhaps there's an event emitted later on that can be piggy backed on? Or would `setImmediate` work here, since the other setImmediate is probably enqueued first?)

Reply:
The `destroy` method calls `finishSessionDestroy` in a `setImmediate` at https://github.com/nodejs/node/blob/e399abd5823b5428e3f3752b3f16f75c302ae471/lib/internal/http2/core.js#L974

While writing tests, I've to wait for `finishSessionDestroy` to complete and setTimeout is called after setImmediate ([stackoverflow discussion](https://stackoverflow.com/a/24119936))

I replaced `setTimeout` with `setImmediate` and the test was successful. I'l updated the commit.